### PR TITLE
chore: RFC for improving S3's support for large uploads

### LINKED ISSUE #3829: Better support for large S3 batches
The current iteration of our S3 sink is built for sending small-to-medium size objects relatively frequently. This is apparent in a few aspects of the config:

* Default max batch size of just 1k events
* Default batch timeout of just 5 minutes
* No way to set an unlimited batch size

This is in contrast with some other tools like [fluentd](https://docs.fluentd.org/output/s3) that will create one file per _hour_ by default. While the different approaches have their pros and cons, it would be a benefit if Vector smoothly supported the type of behavior users may be accustomed to from those other tools.

While users could technically just increase Vector's batch sizes and timeouts to approach the desired behavior, the implementation is not designed for it and it's not clear how well it would perform. There are a few things we could explore doing that would potentially help performance and resource usage:

* Disk-based batching
* Multi-part uploads

These each add a significant degree of complexity (mostly around state management across restarts), but they would decouple memory use from the desired S3 object size and likely provide smoother operation.

### PR DESCRIPTION
**Pull request by bruceg**
Closes #3829 

[Rendered view](https://github.com/vectordotdev/vector/blob/s3-large-upload-support-rfc/rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md)

### CODE REVIEW

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+The following upload behaviors are possible through the configuration
+described below:
+
+1. The sink operates as before, where batches are buffered in memory
+   and uploaded directly to objects. This will initially be the
+   default configuration to avoid breaking existing setups.
+1. The sink buffers batches until it can generate a complete part, and
+   assembles the parts into an object only at the end of an
+   interval. None of the intermediate parts are visible until
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: Looks like this sentence wasn't finished. And to clarify, this is local batching on Vector's host? Data doesn't reach S3 until the end of the interval?

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+   end of an interval _or_ when restarting the upload would
+   not be possible due to an undersized part.
+
+This change will introduce the following new configuration items for
+this sink, which will all default to operating the same as the
+existing sink.
+
+- `data_dir` (string): The path to the directory in which the state
+  data is stored. If this is not set, it is derived from the global
+  `data_dir` setting.
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:116]: Do we have precedence for overriding this? I don't think it's necessary.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+  used. In addition to common interval specifications (ie `"15
+  minutes"`, etc), the following values are permitted:
+  - `"minutely"`: Complete the upload at the end of the minute exactly.
+  - `"hourly"`: Complete the upload at the end of the hour exactly.
+  - `"daily"`: Complete the upload at midnight each day.
+- `upload_incomplete` (boolean): Controls if the sink will upload
+  parts and assemble the upload on shutdown. This causes data to be
+  persisted to S3 across restarts, but will cause the creation of
+  multiple objects where normally there would only be one. Defaults to
+  `false`.
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:133]: I'm wondering if these options are more obvious as:

* `multipart.interval` (enabled if set, default `null`)
* `multipart.shutdown = "upload"` (or something more obvious)

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+unclear, however, if this behavior will interact badly with object
+tracking sources like SQS, or if AWS would penalize this action
+pattern for high volume uploads.
+
+## Outstanding Questions
+
+- Should incomplete parts be unconditionally buffered to disk, to
+  allow for producing parts larger than would be feasible in memory,
+  or buffered in memory as usual and only saved to disk to handle
+  persistence across reloads?
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:357]: I think so. This also unblocks #11662, since we need to disk buffers to realistically enable e2e ack'ing. Performance would be roughly the same, right?

**Review by binarylogic — CHANGES_REQUESTED** [review]
Looking good! I had a few important questions. Let me know if it's easier to chat.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 2 above)
Comment (spencergilbert) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:116]: It's possible on a number of components already 👍

Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:116]: The `file`, `journald`, and `kubernetes_logs` sources all have a `data_dir` setting to allow such an override.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 3 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:133]: Perhaps. I don't have a strong opinion. I'm not excited about having a setting that makes such a major change to the behavior in a sub-config, but I'm open to other considerations.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+
+One of the actions possible when creating a multipart object is to
+copy data from an existing object. This can be used to both copy data
+from and replace an object at the same time, allowing for effective
+append (or prepend) behavior without using extra bandwidth. It is
+unclear, however, if this behavior will interact badly with object
+tracking sources like SQS, or if AWS would penalize this action
+pattern for high volume uploads.
+
+## Outstanding Questions
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:352]: The upload structure and parts still live in the bucket until they are either purged by the bucket lifecycle configuration or are assembled into an object, which can be done after the fact. They are just not visible as normal objects.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+    /// A list of uploaded parts e_tags. The part number is derived from the index.
+    parts: Vec<String>,
+}
+```
+
+## Drawbacks
+
+- By increasing the time over which events are written to the sink, we
+  are increasing the chances of data loss due to crashes and network
+  interruptions.
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:300]: Hmm, yes, that is less than clear. I think I was referring to a different approach in an earlier draft. I will try to clarify.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 3 above)
Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:133]: I'm not either. I feel like there is an opportunity to apply the [be opinionated principle](https://github.com/vectordotdev/vector/blob/master/docs/USER_EXPERIENCE_DESIGN.md#be-opinionated--reduce-decisions) here. I would much rather have one way of doing things if it's better for the user. I'm having a hard time seeing why a user would not want to adopt multipart uploading, especially if we upload that parts as visible objects that get reassembled later.

1. It improves durability by getting data onto a durable storage sooner.
2. It reduces cost by consolidating data on the end storage.
3. It improves in-place query times by bringing files closer to the recommended size for Athena (~250mb).

We should consider making this the only way. cc @jszwedko.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 4 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:357]: I hadn't thought of the persistent store angle. That's a good point, and it's actually mostly independent of the multipart upload changes.

It would decrease performance slightly, as events would have to be written to disk and then reloaded. For any performance sensitive situation, that data would still be in cache on reload, and may not even have to be flushed to disk by the kernel if we manage it right, but there would still be a small overhead. Since we would be encoding events before dumping to disk, this would not cause any serialization overhead, though.

Ideally, we would also want to change the batch interface to allow for streaming data from a batch, as the current interface assumes everything is purely in-memory. Streaming data asynchronously is possible with hyper, and would be an ideal use case for this. For a 5MB batch, however, it's probably not worth the effort.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:352]: Yeah, that's my concern. Most S3 buckets have a life cycle rule that'll delete stale multipart uploads, which would result in data loss. So, to me, this violates the guarantee that e2e ack'ing implies. The data isn't actually written. This makes me want to upload the parts as regular objects that get assembled later. This has the advantage of:

1. Making the data immediately available for analysis with tools like Athena.
2. Prevents the case above where the data gets deleted or lost.

To prevent data loss, how would someone practically reassembled orphaned parts? I've never seen a tool that does this since it's usually the responsibilities of the tool uploading the data.

[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:356]: There are API calls to list current multipart uploads and the associated uploaded parts. The orphaned parts could then be reassembled from this data.

Another downside to uploading as regular parts is that the data would show up twice to anything that is monitoring the uploads in real time (since the events would show both in the individual files and in the assembled larger object).

[the remark this answers is not in the document]
Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:356]: I see. Then I'd like to think about how Vector could solve this in the event of a Vector process crash.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 4 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:361]: On the other hand, it's worth asking what is the real difference between an in-sink disk batch buffer and a pre-sink disk buffer? If we aren't going to be uploading parts that are larger than memory (and 5MB is probably not a big concern here), we have an existing mechanism that has at least a similar effect.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:356]: I should have clarified earlier: The multipart upload identifiers and uploaded part information will necessarily be written to a checkpoint (needed to handle non-crash reloads and restarts as well) so one of the startup tasks will have to be to check if any of those uploads are past their completion time and assemble them.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 4 above)
Comment (lukesteensen) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:361]: It does feel like there's a lot of potential conceptual overlap with this and disk buffers. It might be worth laying out explicitly a breakdown of concerns for each. I'm also curious if there's some potential generalization of the existing disk buffer implementation that could accomplish what we're talking about here.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukesteensen) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:356]: Generally I think it's a good practice to assume at startup that we may have crashed in the middle of an operation and do these kinds of checks to try to rebuild our state as much as possible. It might even be worth investing some time in a subcommand or similar that can inspect this state on disk without needing to start up Vector with a matching config.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+- `batch.min_bytes` (integer): This specifies the target size a
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (duration or enum): How often should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. In addition to common interval specifications (ie `"15
+  minutes"`, etc), the following values are permitted:
+  - `"minutely"`: Complete the upload at the end of the minute exactly.
```
Comment (lukesteensen) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: This doesn't seem correct from an English perspective. I wonder if it'd be clearer to change these to something that's more explicit about it indicating a particular position in each minute/day/hour.

Comment (spencergilbert) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: Somewhat obvious answer of cron style syntax - though that definitely covers more than we need

Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: Alternatively we stick with our current pattern of specifying intervals in seconds.

**Comment by bruceg** [comment]
Per our discussion, I will be reworking this RFC to:
1. Drop the plan to produce persistence in the in-sink batch buffers, instead relying on the existence of a pre-sink disk buffer to handle that capability.
2. Change the normal behavior of this sink to always use multi-part upload, with an optimization to do the normal single-part upload when possible.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (string): When should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. the following values are permitted:
+  - `"hourly"`: Complete the upload at the end of the hour exactly.
+  - `"daily"`: Complete the upload at midnight local time each day.
+  - `"at TIME"`: Complete the upload at a specific time each day.
```
Comment (spencergilbert) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: This still feels like an awkward configuration section - I'd like to see if we could improve the UX here (though I did miss the meeting about this RFC, so I don't know if this was discussed)

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+with the additional behavior change of encoding events before batching
+rather than after.
+
+#### Add a minimum batch size limit to the batcher
+
+Since the multipart object creation API requires each part to have a
+minimum size, this batcher will also have to enforce a minimum size
+before an upload is initiated. The `Batcher` interface will be
+modified to respect this interface and not emit batches until one is
+ready (except at shutdown).
```
Comment (spencergilbert) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:171]: Would we be looking to "complete" the object when shutting down? My understanding from what you stated above only the "last" batch could be under the min size.

**Review by spencergilbert — APPROVED** [review]
I broadly approve, and had one question - as well as a non-blocking UX _feel_ about the interval configuration.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 8 above)
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: I'm not loving this option. A few reasons:

1. I think most users will want to base multipart finalization on an overall target size (i.e., 100mb) as opposed to a time range.
2. It's not clear if this time range is on the hour or within an hour of starting the multipart job.
3. It's not clear that this enables a multipart upload process that results in a final large object.

I think it may be better to provide settings that address the final object and follow the same pattern as `batch`. Call it `object`:

* `object.max_bytes` - Maximum number of bytes for the final object.
* `object.timeout_secs` - Object is finalized after this time regardless of the size.

What do you think?

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Set up the new partitioned batcher.
+- [ ] Initial MVP implementation in S3 with only the simplest upload
+      interval and no persistent buffers.
+- [ ] Add support for all upload interval types.
+
+## Future Improvements
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:320]: I'd like to also mention improvements to object finalization. What happens if a Vector instance crashes and never comes back? How will that object be finalized?

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 8 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: 1. I was taking a cue from the [fluent s3 plugin](https://docs.fluentd.org/output/s3) which outputs objects hourly, albeit using local buffering instead of multipart uploads. It has a [timekey specifier](https://docs.fluentd.org/configuration/buffer-section#time) that controls this in a rather different way. If, however, an object size is the most desired configuration, I'll adjust for that.
2. Fair point. I had intended for `hourly` to mean on the hour.
3. The use of multipart vs single part should be transparent to the user. Essentially, the sink will upload parts _when it can_, and then create the object at the specified point (either time or size based).

I think having a specification for the final object size/time separate from the batch makes sense. I'm a little leery of the specific semantics of the settings described above, but the structure works.
1. Should the size limit be a hard maximum or a "close enough" maximum (ie if it goes one event over, that's OK)?
2. Is creating an object at the time of the first event + T (ie timeout-like behavior) more useful than fixed intervals (such as at exactly 00:00 every hour)?

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+- Support for large batches in the Azure Blob, Datadog Archive, and
+  GCP Cloud Storage sinks.
+
+## Terminology
+
+- Batch: A sequence of events collected within vector in preparation
+  for uploading to a storage sink.
+- Object: A completed upload stored on the destination sink's server(s).
+- Part: A completed partial upload that will be composed into an object.
```
Comment (jszwedko) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:26]: 👍 on this terminology section.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 8 above)
Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: My preference would be to keep the batch configuration as-is, that is if I configure:

```toml
batch.max_bytes = 1073741824 # 1 GB
batch.max_events = 100_000
batch.timeout_secs = 9000
```

Then objects will be finalized from their parts when either the sum of the size of the parts would exceed the max, the sum of the number of events in the parts would exceed 100,000,  or 9000 seconds has passed.

We then add new configuration for the multi-part uploads like:

```toml
batch.min_part_bytes = 5242880 # 5 MB
```

Where Vector would attempt to build parts of at least this size to comprise the full batch.

If `min_part_bytes` exceeds `max_bytes` then we we do not do any partial uploads and the sink behaves as it does currently.

I think this would keep the batching behavior roughly in-line with other sinks.

I do agree with you, for the blob sinks, which typically include time specifiers in their partitioning it might make sense to augment them to be able to run "on the hour" or similar, but consider that outside of the scope of this RFC. It's something we could follow up on separately.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (string): When should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. the following values are permitted:
+  - `"hourly"`: Complete the upload at the end of the hour exactly.
+  - `"daily"`: Complete the upload at midnight local time each day.
+  - `"at TIME"`: Complete the upload at a specific time each day.
+
+### Implementation
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:130]: Agreed, I would like to see that discussed a bit more here. Another consideration, what if Vector is restarted with a changed partitioning scheme?

**Review by jszwedko — COMMENTED** [review]
Thanks for the updates! I left a couple of notes. I agree with Ben that I think we need to add more detail here about how object finalization happens in the face of Vector restarts and crashes.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 8 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: Yes, this is a good simplification to reduce the scope of this effort. I will adjust accordingly.

Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: I'm ok with @jszwedko's proposal, but I do think it is slightly misleading. The finalization steps that reassembles the parts is entirely new and unique to S3. To me, it's completely separate from the batch+request process that exists today. Because of this, I think it deserves a new set of options. It's also worth thinking about doing this for other sink (i.e., the `file` sink) and what the options might look like.

Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:128]: I agree with @binarylogic here about the configuration, but it leads to a bit of a conflict. If we add an `object.size` to control the desired size of the resulting object, how does that functionally differ from what `batch.max_bytes` does now? Should we deprecate and remove the latter?

**Review by jszwedko — COMMENTED** [review]
Thanks for the updates! I'm still wondering a bit about the specifics of recovery (as mentioned below).

@binarylogic I couldn't figure out how to comment [on the thread about the configuration](https://github.com/vectordotdev/vector/pull/11835#discussion_r828156148) so commenting here instead. The way I view it, having the `batch` configuration correspond to the creation of an S3 object as this is consistent with other sinks where a "batch" corresponds to a full "payload" (e.g. HTTP request body, Kafka message, or, in this case, an S3 object). For the `file` sink, if batching was supported, I think would correspond to a full file (which we would write to in "parts"). What we are doing here is uploading a batch in parts which I think goes well with `min_part_size`.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+
+- `batch.min_part_bytes` (integer): This specifies the target size a
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set. If this is
+  larger or equal to `batch.max_bytes`, the batches will always be
+  uploaded as single objects.
+- `recover_partials` (bool): When set, vector will scan for unfinished
+  objects at startup and reassemble them. These are multipart uploads
+  that are interrupted by a crash.
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:117]: Hmm, I thought I had included more verbiage on it. I will expand the relevant section in the proposal above.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+with the additional behavior change of encoding events before batching
+rather than after.
+
+#### Add a minimum batch size limit to the batcher
+
+Since the multipart object creation API requires each part to have a
+minimum size, this batcher will also have to enforce a minimum size
+before an upload is initiated. The `Batcher` interface will be
+modified to respect this interface and not emit batches until one is
+ready (except at shutdown).
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:160]: The final batch can be smaller than the minimum. Since we are going to immediately assemble an object at shutdown, that works.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+1. The existing incomplete uploads are listed with the
+   `ListMultipartUploads` action.
+2. Included in the listing are the object names the uploads would be
+   written to. The `key_prefix` in the configuration is translated
+   into a pattern, and these object names are filtered to match only
+   those that could have come from the current configuration.
+3. For each resulting multipart upload, the parts that compose it are
+   listed with one or more `ListParts` actions.
+4. The upload assembled to the final object with the
+   `CompleteMultipartUpload` action as above.
```
Comment (binarylogic) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: Just verifying, is this operation sensitive to race conditions? What if multiple Vector instances all do this at once?

Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: If multiple Vector instances are configured to write to the same bucket with the same `key_prefix`, then yes one could reassemble uploads another is still working on. I can't think of a good way to avoid this other than either persisting the objects the current system is working on (which I understood to be undesirable, and wouldn't handle the case of a crashed node that never comes back) or manually configuring the recovery to only happen on one instance. Possibly a timeout could work (ie only reassemble uploads older than some threshold), but that too has issues if there is a slow writer.

Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: I think it is fine if another Vector instance finalizes the upload. My question, similar to Ben's, is what would happen if two Vector instances both tried to finalize the upload? My guess is that one of them would fail because the multi part upload was completed by another instance. I think this is fine. We can just output a warning in that case.

Another case to consider is Vector A starting up and finding a partial upload started by Vector B (that is still running). Assuming that Vector A finalizes B's batch, how would B be affected?

One more edge case, what if Vector crashes and then is started with a new prefix?

I also think we'll want to clearly document this lest people suffer data loss because Vector crashed, but they never restarted it.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+  data in the S3 bucket due to incomplete multipart uploads. This data
+  is recoverable, but only with an extra processing step.
+
+## Prior Art
+
+The [S3 output plugin](https://docs.fluentd.org/output/s3) for fluentd
+uses a disk buffer to store the object before uploading, but uses the
+standard `PutObject` action to upload the object.
+
+## Alternatives
```
Comment (jszwedko) [inline rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:253]: I think one other possible alternative that is just a slight twist on the above is to use the AWS S3 bucket to track state by writing some metadata object to track partial uploads. This avoids needing to deal with storing Vector state while solving some of the issues mentioned above around scanning for in-flight partial uploads.

**Review by jszwedko — COMMENTED** [review]
A couple of last comments, otherwise 👍

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 15 above)
Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: Agree, I'm ok if one of the instances fails assuming the operation succeeds on another. We could have thoughtful retries too. That is, if an operation fails, check if the operation is still needed before retrying. Finally, this is a consideration for central state management and coordination (AKA the enterprise product).

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 16 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:253]: What additional data would be stored in this external state file than what is already present in the multipart upload state that S3 already provides?

Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:253]: Ah, I was thinking of whatever you'd originally planned to include in Vector's local `data_dir` to resume uploads. A sort of primitive form of remote state.

Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:253]: Yeah, the state file would pretty much be just what S3 provides with the difference that, since it is local to the node, it would allow for distinguishing between different node's in-progress uploads and avoid clobbering as above. We could store this on S3 with a local node unique identifier, but then we need to store that identifier locally… and still need the `data_dir`.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 15 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: If two Vector instances try to finalize the same object, I think only one would succeed. It should be very difficult to test this, though, as it is a race condition requiring the second request to come in before AWS S3 finishes its operation.

The second edge case is the more troubling one. In that case, Vector A could trash B's upload. Hmm, although in that case the data is still there on S3, B would just need to start a new upload and continue. I'll add that to the process description.

If Vector crashes and is restarted with a new prefix, and no other instance shares the same prefix, then the upload would never be reassembled.

Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: We should peruse the docs and verify this won't have conflicts, because we raise the chances significantly in scenarios where many Vector instances boot at once.

**Review by jszwedko — APPROVED** [review]
Looks good, thanks @bruceg ! My main request is that we clearly document how the multi-part upload behavior recovers from crashes: explicitly noting that Vector must be restarted with the same key prefix after a crash in order to materialize the partially uploaded data.

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md (hunk 15 above)
Comment (bruceg) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: I tested the handling of adding to a deleted multipart upload, and it produces a HTTP 404 as expected, which is something we can catch as described in the latest update. With that process, I don't see a race/conflict possibility that wouldn't have to be handled by S3 itself.

Comment (binarylogic) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:104]: Great, this should be resolved then.

**Review by binarylogic — APPROVED** [review]
Nice work!

File: rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+the implementation decisions below:
+
+1. Vector may need to continue a multipart upload process across sink
+   restarts (due to reloads, restarts, or crashes). Since the uploads
+   require the presence of a unique identifier for both the multipart
+   upload itself as well as each of the uploaded parts, Vector will
+   need to track these identifiers with the partition that initiated
+   the upload.
+2. The upload parts (except for the final part) have a _minimum_ size
+   limit of 5MB. Since the sink may need to be restarted before a
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md:85]: Hey @fptyloun!

That's correct, the batches will still be built in memory. The approach outlined here should reduce memory use per partition to 5 MB, but we are still considering on-disk batching in the future as an enhancement to this approach.
